Add AccessToken to UserCredentials constructors.

anthmgoogle opened the pull request:
3-legged-auth has access token with optional refresh token. Breaking change to UserCredentials constructors.

anthmgoogle commented:
@garrettjonesgoogle and @ejona86 can you review?

garrettjonesgoogle commented on oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java:
```diff
@@ -106,6 +106,22 @@ public static GoogleCredentials fromStream(InputStream credentialsStream, HttpTr
+   **/
+  public GoogleCredentials() {
+    this(null);
+  }
```
Generally I've seen constructors come at the top of a class.

anthmgoogle replied:
The prevailing convention from the Java V1 libraries was to have them after factories.

garrettjonesgoogle replied:
Ahh ok, this seems fine then.

garrettjonesgoogle commented on oauth2_http/java/com/google/auth/oauth2/UserCredentials.java:
```diff
@@ -98,6 +103,10 @@ static UserCredentials fromJson(Map<String, Object> json, HttpTransport transpor
   public AccessToken refreshAccessToken() throws IOException {
+    if (refreshToken ==  null) {
+      throw new IOException("UserCredentials instance cannot refresh because no refresh token was"
+          +" specified in constructor.");
```
add a space after '+'. Also, you might want to avoid referring to coincidental details of the implementation, such as the fact that currently, the only path to providing the refresh token is the constructor; maybe we might add other paths to setting it later.

anthmgoogle replied:
Agreed. Changed.

garrettjonesgoogle commented on oauth2_http/java/com/google/auth/oauth2/UserCredentials.java:
```diff
@@ -36,39 +36,44 @@
   public UserCredentials(String clientId, String clientSecret, String refreshToken,
-      HttpTransport transport, GenericUrl tokenServerUrl) {
+      AccessToken accessToken, HttpTransport transport, GenericUrl tokenServerUrl) {
+    super(accessToken);
```
Adding new parameters in the middle of a parameter list is potentially hazardous, especially when nulls are being thrown around.

anthmgoogle replied:
Agreed there is some risk here. Before locking down the AcessToken definitely belongs before the HttpTransport and GenericUrl, as the latter are only for testing or relatively rare extensible scenarios, while AccessToken is for the most common real scenario. We could have a discussion about whether the breaking change should be staged in some way. I'm inclined to go with this since the project is in alpha and it will break cleanly since the types are all different.

garrettjonesgoogle replied:
Ok it seems like it should be fine since the project is in alpha.

garrettjonesgoogle commented on oauth2_http/javatests/com/google/auth/oauth2/UserCredentialsTest.java:
```diff
@@ -32,6 +33,15 @@
+      fail("Should not be able to refresh without refresh token.");
+    } catch (Exception expected) {
+    }
+  }
```
Do we not have access to the version of junit that supports @Test(expected=Exception) ?

anthmgoogle replied:
We do have that one. Changed this case. I left the case below but left it with the try/catch since there are multiple statements in the test that could cause it.

anthmgoogle commented:
Addressed code review feedback. One behavior change was to change the exception from refreshing the credentials with no refresh token into IllegalStateException which seems like a better fit. I can squash commits we are ready to submit.

garrettjonesgoogle commented:
LGTM.

ejona86 commented on oauth2_http/java/com/google/auth/oauth2/UserCredentials.java:
```diff
@@ -36,39 +36,44 @@
     this.clientId = Preconditions.checkNotNull(clientId);
     this.clientSecret = Preconditions.checkNotNull(clientSecret);
-    this.refreshToken = Preconditions.checkNotNull(refreshToken);
+    this.refreshToken = refreshToken;
```
Hmm... aren't the clientId and clientSecret unnecessary if you only have an access token (no refresh token)? Should there be a 1-arg constructor that just accepts an AccessToken?

anthmgoogle replied:
Yes, you can theoretically use a user access token without a client id or secret. I think to enable this, we want a different class, since if you only have an access token, then it is a general purpose temporary credential. To model this I think we should either have a TemporaryCredentials class, or we should change OAuth2Credenthals to be a non-abstract type where you can pass in an AccesToken directly. I was planning to do this in a subsequent change. But for 3LO scenarios, even without a refresh token, you would always have the Client ID and secret on hand because you need it get the access token.

ejona86 replied:
Okay, that works for me.

ejona86 commented:
@anthmgoogle, FYI, the `UserCredentials` constructor change is no problem for grpc, as we aren't using the class directly.

ejona86 commented:
@anthmgoogle LGTM
